crates/fuel-core/src/state/historical_rocksdb.rs

```diff
@@ -246,41 +275,133 @@ where
     fn rollback_last_block(&self) -> StorageResult<u64> {
-        let latest_height = self
-            .db
-            .iter_all_keys::<ModificationsHistory<Description>>(Some(
-                IterDirection::Reverse,
-            ))
-            .next()
-            .ok_or(DatabaseError::ReachedEndOfHistory)??;
+        let modifications_history_migration_in_progress = self.is_migration_in_progress();
+
+        let (v2_latest_height, v1_latest_height) = self.multiversion_changes_heights(
+            IterDirection::Reverse,
+            modifications_history_migration_in_progress,
+        );
+
+        let latest_height = match (v2_latest_height, v1_latest_height) {
+            (None, None) => Err(DatabaseError::ReachedEndOfHistory)?,
+            (Some(Ok(v1)), Some(Ok(v2))) => v1.min(v2),
+            (_, Some(v1_res)) => v1_res?,
+            (Some(v2_res), _) => v2_res?,
+        };
 
         self.rollback_block_to(latest_height)?;
 
         Ok(latest_height)
     }
 
+    fn take_migration_changes(&self) -> (Changes, Option<ParkingMutexGuard<Changes>>) {
+        if self.is_migration_in_progress() {
+            let mut lock_guard = self.migration_changes.lock();
+            (std::mem::take(&mut *lock_guard), Some(lock_guard))
+        } else {
+            (Changes::default(), None)
+        }
+    }
+
+    fn add_migration_changes(
+        &self,
+        lock_guard: Option<ParkingMutexGuard<Changes>>,
+        changes: Changes,
+    ) -> StorageResult<()> {
+        match lock_guard {
+            // The only case when we did not acquire a lock_guard is because the migration is not in progress.
+            // We do not need to do anything in this case.
+            None => Ok(()),
+            Some(mut lock_guard) => {
+                let current_changes = std::mem::take(&mut *lock_guard);
+                let mut current_changes_transaction = StorageTransaction::transaction(
+                    &self.db,
+                    ConflictPolicy::Overwrite,
+                    current_changes,
+                );
+
+                StorageTransaction::transaction(
+                    &mut current_changes_transaction,
+                    ConflictPolicy::Overwrite,
+                    changes,
+                )
+                .commit()?;
+
+                *lock_guard = current_changes_transaction.into_changes();
+
+                Ok(())
+            }
+        }
+    }
+
     fn rollback_block_to(&self, height_to_rollback: u64) -> StorageResult<()> {
-        let mut storage_transaction = self.db.read_transaction();
+        let (cumulative_changes, maybe_cumulative_changes_guard) =
+            self.take_migration_changes();
+
+        // Will clone an empty set of changes if the migration is not in progress, which should
+        // not impact performance. However, when a migration is in progress, this operation could
+        // be expensive if many changes have been accumulated.
```

Comment:
Short answer: it is possible, but the logic becomes complicated and I'm happy with keeping the lock for a considerably long time. 

Long answer: At a logical level, we have three kind of transactions: "rollback", "commit" and "migrate". The first two commit all the cumulative changes from memory into the underlying rocksdb, while the latter only adds its changes the cumulative changes in memory.  

We can technically keep the lock guard for a short period of time, only when taking and pushing cumulative changes in the shared memory changes, but there is some extra care to be taken: 
* Changes from "commit" and "rollback" transactions are considered to be newer, hence they should take priority and overwrite changes from migraiton transactions. When we want to add new changes to the in-memory shared cumulative changes, we simply overwrite them on top of any change that we might have. 
* Migration changes can be computed independently, but once we want to commit them we will need to acquire the lock on the shared portion of cumulative changes, create a new transaction with the cumulative changes as base, and commit the migration changes on top of it. We must make sure that the migration changes do not overwrite any "newer" changes from other transactions, so a `ConflictPolicy = Fail` must be used for migration transactions. Even if the transaction fails, we know that another transaction has updated the key to a newer value (which will be stored as V2). 
* There is still a chance of a race where the migration changes of a key are written to the shared memory after newer V2 data for that key has been committed to the DB. In this case the migration transaction will succeed despite having a `ConflictPolicy = Fail`. If we want to avoid that altogether, we must either keep the lock while changes are flushed to the database (which basically invalidates any performance benefit), or we must keep the cumulative changes for longer after they have been committed to RocksDB, to prevent migrating transactions from committing stale data. This now poses the problem to have to remove cumulative changes from memory, which I'm not 100% how to do safely.

Reply:
(Btw, another way around it is to use a `RwLock`. Migrating transactions can acquire a `Read` lock, and other transactions can acquire a `Write` lock. But I'm afraid having a read lock for something that writes to memory might raise some eyebrows)

Reply:
Thanks for elaboration. I think if we can live with the lock being locked for some longer time, there's no need to complicate the code even more.